feat: (macOS) Add support for PasswordBox

ajpinedam opened the pull request:
GitHub Issue (If applicable): #
closes: #3005 

## PR Type
- Feature

## What is the current behavior?
PasswordBox is not supported on macOS

## What is the new behavior?
macOS does support PasswordBox

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:
```diff
@@ -44,6 +44,7 @@ public partial class TextBox : Control, IFrameworkTemplatePoolAware
 
 		private ContentPresenter _header;
 		private bool _isPassword;
+		protected bool _isButtonEnabled = true;
```
Make it a `private protected` property instead.

jeromelaban replied:
Make this a `protected private` property.

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -17,6 +17,9 @@ public partial class PasswordBox : TextBox
 
 		public PasswordBox()
+#if __MACOS__
+			:  base(true)
```
Looks like this constructor in TextBox is not used at all, we can probably remove it completely.

ajpinedam replied:
Removed and added only on .macOS partial

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -130,5 +133,43 @@ public override string GetAccessibilityInnerText()
+			get => (bool)this.GetValue(IsPasswordRevealButtonEnabledProperty);
+			set
+			{
+				this.SetValue(IsPasswordRevealButtonEnabledProperty, value);
```
Lambda setter

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -17,6 +17,9 @@ public partial class PasswordBox : TextBox
 
 		public PasswordBox()
+#if __MACOS__
+			:  base(true)
```
it does not look like this TextBox is used in any platform. You can remove it from TextBox completely.

davidjohnoliver commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -51,6 +54,10 @@ private void RegisterSetPasswordScope()
 				});
+
+#if __MACOS__
+				IsPasswordRevealButtonEnabled = false;
```
I would put this in the Style, along with a comment on why it's needed

davidjohnoliver commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -130,5 +133,43 @@ public override string GetAccessibilityInnerText()
+			UpdateButtonStates();
+		}
+
+		private void UpdateButtonStates()
```
It's kinda confusing that this duplicates the functionality of a different method with the same name in `TextBox`, with no sync-up between the two

ajpinedam replied:
Removed method since it was not being used anywhere else and added functionality to the caller.

davidjohnoliver commented on src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/PasswordBox.cs:
```diff
@@ -23,7 +23,7 @@ public  string PasswordChar
 		// Skipping already declared property Password
 		// Skipping already declared property MaxLength
-		#if __ANDROID__ || __IOS__ || NET461 || __WASM__ || __MACOS__
+		#if __ANDROID__ || __IOS__ || NET461 || __WASM__ || false
```
I think this should be false for all platforms? It seems like you've implemented it for all of them

ajpinedam commented:
![macOS_Login](https://user-images.githubusercontent.com/1900897/80025654-fe122780-84ae-11ea-960e-868bbe918be7.gif)

jeromelaban commented:
@ajpinedam looks like [the build failed](https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=14168&view=logs&j=b0e9e703-bc7b-5b6a-2ec3-61742e5311eb&t=fe694602-11e4-517a-11ac-0f0951e3424d&l=6358).

nventive-devops commented on src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.macOS.cs:
```diff
@@ -1,24 +1,23 @@
 {
 	public partial class TextBox
 	{
+		private bool _isPassword;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make '_isPassword' 'readonly'.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5398636)

ajpinedam commented:
The build is failing with this error, what is it about?

```
##[error]EXEC(0,0): Error : Removed method Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.PasswordBox.get_IsPasswordRevealButtonEnabledProperty() not found in ignore set.

##[error]EXEC(0,0): Error : Removed method System.Void Windows.UI.Xaml.Controls.TextBox..ctor(System.Boolean isPassword) not found in ignore set.

##[error]EXEC(0,0): Error : Removed property Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.PasswordBox::IsPasswordRevealButtonEnabledProperty() not found in ignore set.
```

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs:
```diff
@@ -130,5 +129,42 @@ public override string GetAccessibilityInnerText()
+			set => this.SetValue(IsPasswordRevealButtonEnabledProperty, value);			
+		}
+
+		public static readonly DependencyProperty IsPasswordRevealButtonEnabledProperty =
```
The error happens because you changed the signature of `IsPasswordRevealButtonEnabledProperty`. It should be: 
```suggestion
		public static global::Windows.UI.Xaml.DependencyProperty IsPasswordRevealButtonEnabledProperty { get; } = 
```

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:
```diff
@@ -71,18 +70,12 @@ public partial class TextBox : Control, IFrameworkTemplatePoolAware
 
 		private void OnParentChanged(object instance, object key, DependencyObjectParentChangedEventArgs args) => UpdateFontPartial();
 
-		protected TextBox(bool isPassword)
```
Since this one is not used, you'll need to add `System.Void Windows.UI.Xaml.Controls.TextBox..ctor(System.Boolean isPassword)` to the [DiffIgnore file](https://github.com/unoplatform/uno/blob/master/build/PackageDiffIgnore.xml). There's a block missing since the update to 2.2.0, carl's PR will add it.
